File: service/internal/health/healthchecker.go
package health

import "context"

// Healthchecker represents any component that can report on it's own health.
type Healthchecker interface {
	// CheckHealth checks the health of the component and returns either nil if the component
	// is healthy or else an error representing the reason the component is unhealthy.
	CheckHealth(ctx context.Context) error
}

// Function type that can be used as a healthchecker.
type HealthcheckerFunc func(ctx context.Context) error

func (f HealthcheckerFunc) CheckHealth(ctx context.Context) error {
	return f(ctx)
}

File: service/crates/health/src/healthchecker.rs
use async_trait::async_trait;
use std::{boxed::Box, error::Error};

/// Trait that components able to report on their health can implement.
#[async_trait]
pub trait Healthchecker: Send + Sync {
    /// Check the health of the component.
    ///
    /// # Returns
    /// If the component is healthy then returns `Ok(())`.
    /// If the component is unhealthy then returns `Err()` containing the reason why the component is unhealthy.
    async fn check_health(&self) -> Result<(), Box<dyn Error>>;
}
